### DISCUSSION TITLE
Torn w.r.t. “uninterruptible” paragraphs

### DISCUSSION THREAD

**Discussion by rauschma** [jgm/djot#295] [Q&A]
I’m playing with Markdown syntax vs. djot syntax at the moment: In a parser and a book I have written.

I’m torn w.r.t. “uninterruptible” paragraphs – that is, we can only start a new block construct after a paragraph if it is separated by an empty line.

## Pros of uninterruptible paragraphs

* Uninterruptible paragraphs help a lot with hard wrapping.
  * Many people do use hard wrapping (much more than I thought): I did [a small informal survey](https://fosstodon.org/@rauschma/112213750986728170) and 63% of 43 people hard-wrap often or sometimes.

## Cons of uninterruptible paragraphs

* I don’t like how they affect spacing.
* Djot not requiring an empty line before a closing fenced div delimiter kind of goes against its own rules and means mentioning “:::” in a hard-wrapped paragraph can cause problems.

### Spacing in lists

```md
* Must be followed by empty line

  * Subitem 1
  * Subitem 2
```

### Spacing in divs

```md
:::::: wrapper
Outer – must be followed by an empty line

::: wrapped
Inner
:::
::::::
```

In my own content, I’m using the following syntax and I don’t like that I have to insert extra empty lines.

Float:

```md
{!float for="Figure" #logo}
:::
![](img/logo.svg)

--- !caption

This is our company logo.
:::
```

Boxout:

```md
{!boxout icon="icon/tip.svg"}
:::
**This is a special tip**

---

This text describes the tip.
:::
```

## What to do?

I don’t know what to do: Each of the two options has pros and cons.

Vague idea:

* By default, paragraphs are interruptible.
* If there is an empty line with a `+` before a paragraph, it becomes uninterruptible.
* Downside: visual clutter.
* Downside: Editors won’t consider `+` lines to be empty – which is bad for hard wrapping.

Example:

``````md
+
Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium
doloremque laudantium, totam rem aperiam, eaque ipsa quae ab illo inventore
veritatis et quasi architecto beatae vitae dicta sunt explicabo. Nemo enim
ipsam voluptatem quia voluptas sit aspernatur aut odit aut fugit, sed quia
consequuntur magni dolores eos qui ratione voluptatem sequi nesciunt.
+
Neque porro quisquam est, qui dolorem ipsum quia dolor sit amet,
consectetur, adipisci velit, sed quia non numquam eius modi tempora
incidunt ut labore et dolore magnam aliquam quaerat voluptatem. Ut enim ad
minima veniam, quis nostrum exercitationem ullam corporis suscipit
laboriosam, nisi ut aliquid ex ea commodi consequatur?
``````

In lists - option 1:

```md
* +
Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium
doloremque laudantium, totam rem aperiam, eaque ipsa quae ab illo inventore
veritatis et quasi architecto beatae vitae dicta sunt explicabo. Nemo enim
ipsam voluptatem quia voluptas sit aspernatur aut odit aut fugit, sed quia
consequuntur magni dolores eos qui ratione voluptatem sequi nesciunt.

* +
Neque porro quisquam est, qui dolorem ipsum quia dolor sit amet,
consectetur, adipisci velit, sed quia non numquam eius modi tempora
incidunt ut labore et dolore magnam aliquam quaerat voluptatem. Ut enim ad
minima veniam, quis nostrum exercitationem ullam corporis suscipit
laboriosam, nisi ut aliquid ex ea commodi consequatur?
```

In lists - option 2:

```md
*
+
Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium
doloremque laudantium, totam rem aperiam, eaque ipsa quae ab illo inventore
veritatis et quasi architecto beatae vitae dicta sunt explicabo. Nemo enim
ipsam voluptatem quia voluptas sit aspernatur aut odit aut fugit, sed quia
consequuntur magni dolores eos qui ratione voluptatem sequi nesciunt.

*
+
Neque porro quisquam est, qui dolorem ipsum quia dolor sit amet,
consectetur, adipisci velit, sed quia non numquam eius modi tempora
incidunt ut labore et dolore magnam aliquam quaerat voluptatem. Ut enim ad
minima veniam, quis nostrum exercitationem ullam corporis suscipit
laboriosam, nisi ut aliquid ex ea commodi consequatur?
```

**Comment by jgm** [comment]
> mentioning “:::” in a hard-wrapped paragraph can cause problems.

I think that's only a theoretical problem.  Nobody is going to write `:::` in a paragraph -- unless it is quoted in backticks, and then it will be fine.

> By default, paragraphs are interruptible

I think that's the wrong default; people will find they have surprising lists in their document due to hard-wrapping of numbers followed by periods, for example.  If you tell them, "well, you should have preceded the paragraph with a `+`", they will not be impressed.  The `+`'s are just syntactic clutter.

**Reply by rauschma** [reply]
Good points – which indicate a solution:

> people will find they have surprising lists in their document due to hard-wrapping of numbers followed by periods [...]

True. I don’t think there is a way around the empty line in before a nested list. 😢

> Nobody is going to write `:::` in a paragraph

My other issues go away if I allow allow a subset of all block delimiters to interrupt paragraphs – in particular:

* Closing div delimiters
* Thematic breaks (except `---`, see below)

One tricky case remains: The `---` syntax for thematic breaks – which is also used for smart punctuation em dashes. But that one is ambivalent anyway – e.g, is the following a block (=thematic break) or a paragraph (that starts with an em dash)?

```md
---
```
